### PR TITLE
detach: detach instance from asg on termination to prevent head-of-line blocking

### PR DESCRIPTION
**Pull request by igorwwwwwwwwwwwwwwwwwwww**
refs travis-pro/dev#45

### CODE REVIEW

File: lifecycle.go [hunk 1]
```diff
@@ -52,7 +53,16 @@ func handleLifecycleTransition(db repo, log logrus.FieldLogger,
 		return nil
 	}
 
-	err = completeLifecycleAction(action, log, asSvc)
+	if transition == "terminating" && os.Getenv("DETACH_INSTANCE_ON_TERMINATION") == "true" {
```
Comment (meatballhat) [inline lifecycle.go]: This is cool for starters.  I'd like to see this env var integrated with the rest of the configuration bits if possible.

File: lifecycle.go [hunk 2]
```diff
@@ -52,7 +53,16 @@ func handleLifecycleTransition(db repo, log logrus.FieldLogger,
...
+		log = log.WithFields(logrus.Fields{
```
Comment (meatballhat) [inline lifecycle.go]: If a single field, then `log.WithField("detach", true)` 👍

**Review by meatballhat — APPROVED** [review]
🤘 🐢

**Review by meatballhat — APPROVED** [review]
🎸 🐠

**Comment by igorwwwwwwwwwwwwwwwwwwww** [comment]
Looks like this approach has some issues:

```
time="2017-06-13T15:42:24Z" level=error msg="handling lifecycle transition failed" err="ValidationError: The instance i-fafafaf is not in InService or Standby.\n\tstatus code: 400, request id: fefefef instance=i-fafafaf method=POST path="/terminations/i-fafafaf"  
```

The instance asg state is already `Terminate:Wait` by the time we are trying to remove it, at which point it is no longer possible to detach it.

I'm looking into performing the detachment earlier in the process, in the sns callback that handles the lifecycle actions.

**Comment by soulshake** [comment]
I think [step scaling](https://github.com/travis-infrastructure/terraform-config/pull/175) solved the problem this PR was meant to address. <3

**Pull request closed without merging**